boards: st: Add support for STM32H745I-Disco (#72510)
zephyrproject-rtos/zephyr: boards/st/stm32h745i_disco/doc/index.rst

.. _stm32h745i_disco_board:

ST STM32H745I Discovery
###################

Overview
********

> The STM32H745I-DISCO and STM32H750B-DK Discovery kits are complete demonstration and development

Reviewer: The text was taken from ST website, I have removed all H750 references

+-----------+------------+-------------------------------------+
| GPIO      | on-chip    | gpio                                |
+-----------+------------+-------------------------------------+
| RTC       | on-chip    | counter                             |
+-----------+------------+-------------------------------------+
| I2C       | on-chip    | i2c                                 |
+-----------+------------+-------------------------------------+
| PWM       | on-chip    | pwm                                 |
+-----------+------------+-------------------------------------+
| ETHERNET  | on-chip    | ethernet                            |
+-----------+------------+-------------------------------------+
| SPI       | on-chip    | spi                                 |
+-----------+------------+-------------------------------------+
| QSPI NOR  | on-chip    | off-chip flash                      |
+-----------+------------+-------------------------------------+
| FDCAN     | on-chip    | fdcan                               |
+-----------+------------+-------------------------------------+
| FMC       | on-chip    | memc (SDRAM)                        |

> +-----------+------------+-------------------------------------+

Reviewer: Added

two cores. This is done in 3 ways:

- **Compilation**: Clock configuration is only accessible to M7 core. M4 core only
  has access to bus clock activation and deactivation.
- **Static pre-compilation assignment**: Peripherals such as a UART are assigned in
  devicetree before compilation. The user must ensure peripherals are not assigned
  to both cores at the same time.
- **Run time protection**: Interrupt-controller and GPIO configurations could be
  accessed by both cores at run time. Accesses are protected by a hardware semaphore
  to avoid potential concurrent access issues.

Programming and Debugging
*************************

Applications for the ``stm32h745i_disco`` board should be built per core target,
using either ``stm32h745i_disco/stm32h745xx/m7`` or ``stm32h745i_disco/stm32h745xx/m4`` as the target
(see :ref:`build_an_application` and :ref:`application_run` for more details).

.. note::

   If using OpenOCD you will need a recent development version as the last
   official release does not support H7 series and ST-LINK V3 yet.
   Following links may be helpful: `OpenOCD installing Debug Version`_

>    and `OpenOCD installing with ST-LINK V3 support`_

Reviewer 2: I don't think this is still applicable, need to be checked.

Reviewer: The support for H7 was added in [0.11.0](https://github.com/openocd-org/openocd/blob/9ea7f3d647c8ecf6b0f1424002dfc3f4504a162c/NEWS-0.11.0#L63). Should this be removed?

Tested with OpenOCD 0.12.0 by `openocd -f boards/st/stm32h745i_disco/support/openocd.cfg `.

Reviewer 2: > Should this be removed?

Yes

Reviewer: Done

.. note::

   Check if your ST-LINK V3 has newest FW version. It can be done with `STM32CubeIDE`_

Flashing
========

STM32H745I-DISCO board includes an ST-LINK/V3 embedded debug tool interface.

Flashing operation will depend on the target to be flashed and the SoC
option bytes configuration.

By default:

  - CPU0 (Cortex-M7) boot address is set to 0x80000000 (OB: BOOT_CM7_ADD0)
  - CPU1 (Cortex-M4) boot address is set to 0x81000000 (OB: BOOT_CM4_ADD0)

Also, default out of the box board configuration enables CM7 and CM4 boot when
board is powered (Option bytes BCM7 and BCM4 are checked).
In that configuration, Kconfig boot option ``STM32H7_BOOT_CM4_CM7`` should be selected.
Zephyr flash configuration has been set to meet these default settings.

> 

Reviewer 2: ```suggestion

Alternatively, west `STM32CubeProgrammer`_ runner can be used, after installing
it, to flash applications for both cores. The target core is detected automatically.

.. code-block:: console

   $ west flash --runner stm32cubeprogrammer
```

Reviewer: Added

.. zephyr-app-commands::
   :zephyr-app: samples/basic/blinky
   :board: stm32h745i_disco/stm32h745xx/m7
   :goals: build flash

.. note::

   Flashing both M4 and M7 and pushing RESTART button on the board leads
   to LD1 and LD2 flashing simultaneously.

Debugging
=========

You can debug an application in the usual way.  Here is an example for the
:ref:`hello_world` application.

.. zephyr-app-commands::
   :zephyr-app: samples/hello_world
   :board: stm32h745i_disco/stm32h745xx/m7
   :maybe-skip-config:
   :goals: debug

Debugging with west is currently not available on Cortex M4 side.
In order to debug a Zephyr application on Cortex M4 side, you can use
`STM32CubeIDE`_.

.. _STM32H745I-DISCO website:
   https://www.st.com/en/evaluation-tools/stm32h745i-disco.html

.. _STM32H745XI on www.st.com:
   https://www.st.com/en/microcontrollers-microprocessors/stm32h745xi.html

.. _STM32H745xx reference manual:
   https://www.st.com/resource/en/reference_manual/rm0399-stm32h745755-and-stm32h747757-advanced-armbased-32bit-mcus-stmicroelectronics.pdf

.. _STM32H745xx datasheet:
   https://www.st.com/resource/en/datasheet/stm32h745xi.pdf

.. _STM32H745-Disco UM:

>    https://www.st.com/resource/en/user_manual/um2488-discovery-kits-with-stm32h745xi-and-stm32h750xb-mcus-stmicroelectronics.pdf

Reviewer 2: ```suggestion
   https://www.st.com/resource/en/user_manual/um2488-discovery-kits-with-stm32h745xi-and-stm32h750xb-mcus-stmicroelectronics.pdf
   
.. _STM32CubeProgrammer:
   https://www.st.com/en/development-tools/stm32cubeprog.html
```

Reviewer: Added

eMMC, USART, SAI audio DAC stereo with audio jack input and output, MEMS digital microphone, SDRAM,
Quad-SPI flash memory, and RGB interface LCD with capacitive multi-touch panel). ARDUINO® Uno V3
connectors provide easy connection to extension shields or daughterboards for specific applications.

STLINK-V3E is integrated into the board, as an embedded in-circuit debugger and programmer for the
STM32 MCU and the USB Virtual COM port bridge

Key Features

- Arm® Cortex® core-based microcontroller with 2 Mbytes of flash memory and 1 Mbyte of RAM, in a TFBGA240+25 package
- 4.3” RGB interface LCD with touch panel connector
- Ethernet compliant with IEEE-802.3-2002, and PoE
- USB OTG FS
- SAI audio codec
- One ST-MEMS digital microphone
- 2× 512-Mbit Quad-SPI NOR flash memory
- 128-Mbit SDRAM
- 4-Gbyte on-board eMMC
- 1 user and reset push-button
- Fanout daughterboard
- 2× CAN FDs
- Board connectors:

 - USB FS Micro-AB connectors
 - ST-LINK Micro-B USB connector
 - USB power Micro-B connector
 - Ethernet RJ45
 - Stereo headset jack including analog microphone input
 - Audio header for external speakers

>  - Tag‑Connect™ (TAG) 10-pin footprint

Reviewer 2: ```suggestion
 - Tag‑Connect |trade| (TAG) 10-pin footprint
```

In that configuration, Kconfig boot option ``STM32H7_BOOT_CM4_CM7`` should be selected.
Zephyr flash configuration has been set to meet these default settings.

Alternatively, west `STM32CubeProgrammer`_ runner can be used, after installing
it, to flash applications for both cores. The target core is detected automatically.

.. code-block:: console

   $ west flash --runner stm32cubeprogrammer

Flashing an application to STM32H745XI M7 Core
----------------------------------------------
First, connect the STM32H745I-DISCO to your host computer using
the USB port to prepare it for flashing. Then build and flash your application.

Here is an example for the :ref:`hello_world` application.

Run a serial host program to connect with your STM32H745I-DISCO board.

.. code-block:: console

   $ minicom -b 115200 -D /dev/ttyACM0

or use screen:

.. code-block:: console

   $ screen /dev/ttyACM0 115200

Build and flash the application:

.. zephyr-app-commands::
   :zephyr-app: samples/hello_world
   :board: stm32h745i_disco/stm32h745xx/m7
   :goals: build flash

You should see the following message on the console:

.. code-block:: console

   $ Hello World! stm32h745i_disco

.. note::

>   Sometimes, flashing is not working. It is necessary to erase the flash

Reviewer 2: ```suggestion
  Sometimes, flashing does not work properly. It is necessary to erase the flash
```

.. _stm32h745i_disco_board:

> ST STM32H745I Discovery
> ###################

Reviewer 2: You have a CI error: "index.rst:4: WARNING: Title underline too short."

```suggestion
ST STM32H745I Discovery
#######################
```
